File: dart_sample/lib/demos/material/autocomplete_demo_page.dart
import 'package:flutter/material.dart';

class AutocompleteDemoPage extends StatefulWidget {
  const AutocompleteDemoPage({super.key});

  @override
  State<AutocompleteDemoPage> createState() => _AutocompleteDemoPageState();
}

class _AutocompleteDemoPageState extends State<AutocompleteDemoPage> {
  static const List<String> _frameworkTerms = <String>[
    'Widget',
    'Element',
    'RenderObject',
    'BuildContext',
    'StatefulWidget',
    'InheritedWidget',
    'Navigator',
    'Autocomplete',
    'RawAutocomplete',
  ];

  final TextEditingController _materialController = TextEditingController();
  final FocusNode _materialFocusNode = FocusNode();
  final TextEditingController _rawController = TextEditingController();
  final FocusNode _rawFocusNode = FocusNode();
  OptionsViewOpenDirection _openDirection = OptionsViewOpenDirection.down;
  String _materialSelection = 'none';
  String _rawSelection = 'none';

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 12,
        children: <Widget>[
          const Text(
            'Autocomplete + RawAutocomplete',
            style: TextStyle(fontSize: 20, color: Colors.black),
          ),
          const Text(
            'Material defaults and a custom raw field/options view with shared filtering, keyboard highlighting, and anchored direction probes.',
            style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: TextButton(
              onPressed: () => setState(() {
                _openDirection = _nextDirection(_openDirection);
              }),
              child: Text('Open: ${_formatDirection(_openDirection)}'),
            ),
          ),
          const Text(
            'Material Autocomplete',
            style: TextStyle(fontSize: 18, color: Colors.black),
          ),
          Autocomplete<String>(
            optionsBuilder: _filterTerms,
            textEditingController: _materialController,
            focusNode: _materialFocusNode,
            optionsViewOpenDirection: _openDirection,
            optionsMaxHeight: 160,
            onSelected: (String value) {
              setState(() => _materialSelection = value);
            },
          ),
          Text(
            'Selected: $_materialSelection',
            style: const TextStyle(fontSize: 13),
          ),
          const Divider(),
          const Text(
            'RawAutocomplete',
            style: TextStyle(fontSize: 18, color: Colors.black),
          ),
          RawAutocomplete<String>(
            textEditingController: _rawController,
            focusNode: _rawFocusNode,
            optionsBuilder: _filterTerms,
            optionsViewOpenDirection: _openDirection,
            displayStringForOption: (String value) => value,
            fieldViewBuilder:
                (
                  BuildContext context,
                  TextEditingController controller,
                  FocusNode focusNode,
                  VoidCallback onSubmitted,
                ) {
                  return TextField(
                    controller: controller,
                    focusNode: focusNode,
                    decoration: const InputDecoration(
                      labelText: 'Framework concept',
                      hintText: 'Type to filter',
                    ),
                    onSubmitted: (String value) => onSubmitted(),
                  );
                },
            optionsViewBuilder:
                (
                  BuildContext context,
                  AutocompleteOnSelected<String> onSelected,
                  Iterable<String> options,
                ) {
                  final int highlightedIndex = AutocompleteHighlightedOption.of(
                    context,
                  );
                  final List<String> materialized = options.toList();
                  return Material(
                    elevation: 4,
                    borderRadius: BorderRadius.circular(12),
                    clipBehavior: Clip.antiAlias,
                    child: ListView(
                      padding: EdgeInsets.zero,
                      shrinkWrap: true,
                      children: <Widget>[
                        for (
                          int index = 0;
                          index < materialized.length;
                          index += 1
                        )
                          InkWell(
                            onTap: () => onSelected(materialized[index]),
                            child: Container(
                              color: index == highlightedIndex
                                  ? Theme.of(context).focusColor
                                  : null,
                              padding: const EdgeInsets.symmetric(
                                horizontal: 16,
                                vertical: 12,
                              ),
                              child: Text(materialized[index]),
                            ),
                          ),
                      ],
                    ),
                  );
                },
            onSelected: (String value) {
              setState(() => _rawSelection = value);
            },
          ),
          Text(
            'Selected: $_rawSelection',
            style: const TextStyle(fontSize: 13),
          ),
        ],
      ),
    );
  }

  @override
  void dispose() {
    _materialController.dispose();
    _materialFocusNode.dispose();
    _rawController.dispose();
    _rawFocusNode.dispose();
    super.dispose();
  }

  Iterable<String> _filterTerms(TextEditingValue value) {
    final String query = value.text.trim().toLowerCase();
    return _frameworkTerms.where(
      (String term) => query.isEmpty || term.toLowerCase().contains(query),
    );
  }

  static OptionsViewOpenDirection _nextDirection(
    OptionsViewOpenDirection value,
  ) {
    return switch (value) {
      OptionsViewOpenDirection.down => OptionsViewOpenDirection.up,
      OptionsViewOpenDirection.up => OptionsViewOpenDirection.mostSpace,
      OptionsViewOpenDirection.mostSpace => OptionsViewOpenDirection.down,
    };
  }

  static String _formatDirection(OptionsViewOpenDirection value) {
    return switch (value) {
      OptionsViewOpenDirection.down => 'down',
      OptionsViewOpenDirection.up => 'up',
      OptionsViewOpenDirection.mostSpace => 'mostSpace',
    };
  }
}

File: src/Sample/Plumix.Sample/Demos/Material/AutocompleteDemoPage.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/material/autocomplete_demo_page.dart

public sealed class AutocompleteDemoPage : StatefulWidget
{
    public override State CreateState() => new AutocompleteDemoPageState();
}

internal sealed class AutocompleteDemoPageState : State
{
    private static readonly IReadOnlyList<string> FrameworkTerms =
    [
        "Widget",
        "Element",
        "RenderObject",
        "BuildContext",
        "StatefulWidget",
        "InheritedWidget",
        "Navigator",
        "Autocomplete",
        "RawAutocomplete",
    ];

    private readonly TextEditingController _materialController = new();
    private readonly FocusNode _materialFocusNode = new();
    private readonly TextEditingController _rawController = new();
    private readonly FocusNode _rawFocusNode = new();
    private OptionsViewOpenDirection _openDirection = OptionsViewOpenDirection.Down;
    private bool _useMaterial3 = true;
    private string _materialSelection = "none";
    private string _rawSelection = "none";

    public override Widget Build(BuildContext context)
    {
        return new SingleChildScrollView(
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 12,
                children:
                [
                    new Text("Autocomplete + RawAutocomplete", fontSize: 20, color: Colors.Black),
                    new Text(
                        "Material defaults and a route-free raw options portal with shared filtering, keyboard " +
                        "highlighting, M2/M3 surfaces, inherited theme, and anchored direction probes.",
                        fontSize: 14,
                        color: Color.Parse("#8A000000")),
                    new Align(
                        alignment: Alignment.CenterLeft,
                        child: new Row(
                            mainAxisSize: MainAxisSize.Min,
                            children:
                            [
                                new TextButton(
                                    new Text($"Open: {FormatDirection(_openDirection)}"),
                                    () => SetState(() => _openDirection = NextDirection(_openDirection))),
                                new SizedBox(width: 8),
                                new TextButton(
                                    new Text(_useMaterial3 ? "Theme: M3" : "Theme: M2"),
                                    () => SetState(() => _useMaterial3 = !_useMaterial3)),
                            ])),
                    new Text("Material Autocomplete", fontSize: 18, color: Colors.Black),
                    new Theme(
                        new ThemeData(useMaterial3: _useMaterial3),
                        new Autocomplete<string>(
                            optionsBuilder: FilterTerms,
                            textEditingController: _materialController,
                            focusNode: _materialFocusNode,
                            optionsViewOpenDirection: _openDirection,
                            optionsMaxHeight: 160,
                            onSelected: value => SetState(() => _materialSelection = value))),
                    new Text($"Selected: {_materialSelection}", fontSize: 13),
                    new Divider(),
                    new Text("RawAutocomplete", fontSize: 18, color: Colors.Black),
                    new Theme(
                        Theme.Of(context) with { FocusColor = Color.Parse("#243F51B5") },
                        new RawAutocomplete<string>(
                            textEditingController: _rawController,
                            focusNode: _rawFocusNode,
                            optionsBuilder: FilterTerms,
                            optionsViewOpenDirection: _openDirection,
                            displayStringForOption: value => value,
                            fieldViewBuilder: (fieldContext, controller, focusNode, onSubmitted) => new TextField(
                                controller: controller,
                                focusNode: focusNode,
                                decoration: new InputDecoration(
                                    labelText: "Framework concept",
                                    hintText: "Type to filter"),
                                onSubmitted: value => onSubmitted()),
                            optionsViewBuilder: BuildRawOptions,
                            onSelected: value => SetState(() => _rawSelection = value))),
                    new Text($"Selected: {_rawSelection}", fontSize: 13),
                ]));
    }

    public override void Dispose()
    {
        _materialController.Dispose();
        _materialFocusNode.Dispose();
        _rawController.Dispose();
        _rawFocusNode.Dispose();
    }

    private static IEnumerable<string> FilterTerms(TextEditingValue value)
    {
        string query = value.Text.Trim();
        return FrameworkTerms.Where(term => query.Length == 0
            || term.Contains(query, StringComparison.OrdinalIgnoreCase));
    }

    private static Widget BuildRawOptions(
        BuildContext context,
        AutocompleteOnSelected<string> onSelected,
        IEnumerable<string> options)
    {
        int highlightedIndex = AutocompleteHighlightedOption.Of(context);
        string[] materialized = options.ToArray();
        var theme = Theme.Of(context);
        return new DecoratedBox(
            new BoxDecoration(
                Color: theme.CanvasColor,
                Border: Plumix.Rendering.Border.FromBorderSide(new BorderSide(theme.ColorScheme.OutlineVariant)),
                BorderRadius: BorderRadius.Circular(12)),
            child: new ClipRRect(
                BorderRadius.Circular(12),
                child: new ListView(
                    shrinkWrap: true,
                    children: materialized.Select((option, index) =>
                    {
                        Widget content = new Padding(new Thickness(16, 12), new Text(option));
                        if (index == highlightedIndex)
                        {
                            content = new ColoredBox(theme.FocusColor, child: content);
                        }

                        return (Widget)new InkWell(onTap: () => onSelected(option), child: content);
                    }).ToArray())));
    }

    private static OptionsViewOpenDirection NextDirection(OptionsViewOpenDirection value)
    {
        return value switch
        {
            OptionsViewOpenDirection.Down => OptionsViewOpenDirection.Up,
            OptionsViewOpenDirection.Up => OptionsViewOpenDirection.MostSpace,
            _ => OptionsViewOpenDirection.Down,
        };
    }

    private static string FormatDirection(OptionsViewOpenDirection value)
    {
        return value switch
        {
            OptionsViewOpenDirection.Up => "up",
            OptionsViewOpenDirection.MostSpace => "mostSpace",
            _ => "down",
        };
    }
}
